Question-1: Summarize Using Gemini
You are provided with the following paragraph:

📝 Input Paragraph: Over the past decade, education has experienced a significant shift due to rapid advancements in technology. Online learning platforms like Coursera, edX, and Khan Academy have made high-quality education accessible to millions of students worldwide. Virtual classrooms now allow real-time interaction between teachers and learners, breaking down geographical barriers. Artificial intelligence is further enhancing the learning experience through personalized tutoring, automated grading, and adaptive content delivery based on student performance. This shift has empowered students to learn at their own pace, revisit lectures, and explore diverse topics beyond their curriculum. Mobile apps and gamified learning modules are keeping learners engaged in innovative ways. However, this digital transformation also introduces challenges. Not all students have equal access to stable internet, quality devices, or a quiet learning environment at home. Additionally, the lack of face-to-face interaction may lead to feelings of isolation and reduced motivation. As we embrace the future of education, balancing technology with human support and inclusivity remains critical.

✅ Your Task: Use Gemini or OpenAI model ChatModel in LangChain to summarize the above paragraph into exactly 5 bullet points. The bullet points should be concise, clear, and informative.

You may prompt like this:

Summarize the following paragraph into 5 bullet points: <paste paragraph here>

In [ ]:
!pip install langchain langchain-google-genai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.3/53.3 kB 1.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.6/65.6 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 426.6/426.6 kB 15.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 489.1/489.1 kB 32.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.3/233.3 kB 14.5 MB/s eta 0:00:00
  Attempting uninstall: google-auth
    Found existing installation: google-auth 2.43.0
    Uninstalling google-auth-2.43.0:
      Successfully uninstalled google-auth-2.43.0
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 1.2.1
    Uninstalling langchain-core-1.2.1:
      Successfully uninstalled langchain-core-1.2.1
  Attempting uninstall: google-genai
    Found existing installation: google-genai 1.55.0
    Uninstalling google-genai-1.55.0:
      Successfully uninstalled google-genai-1.55.0
ERROR: pip's dependency resolver does not currently take into

In [12]:
#!pip install langchain langchain-google-genai
from langchain_google_genai import ChatGoogleGenerativeAI
from google.colab import userdata
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
llm = ChatGoogleGenerativeAI(model="gemini-2.5-flash-lite",google_api_key=userdata.get('GEMINI_API_KEY'))
ip_para = """
            Over the past decade, education has experienced a significant shift due to rapid advancements in technology.
            Online learning platforms like Coursera, edX, and Khan Academy have made high-quality education accessible to millions of students worldwide.
            Virtual classrooms now allow real-time interaction between teachers and learners, breaking down geographical barriers.
            Artificial intelligence is further enhancing the learning experience through personalized tutoring, automated grading, and adaptive content delivery based on student performance.
            This shift has empowered students to learn at their own pace, revisit lectures, and explore diverse topics beyond their curriculum.
            Mobile apps and gamified learning modules are keeping learners engaged in innovative ways. However, this digital transformation also introduces challenges.
            Not all students have equal access to stable internet, quality devices, or a quiet learning environment at home.
            Additionally, the lack of face-to-face interaction may lead to feelings of isolation and reduced motivation.
            As we embrace the future of education, balancing technology with human support and inclusivity remains critical.
"""
points = ["clean","clear","concise","informative"]
para_prompt= PromptTemplate(input_variables=["para","points"],template="Summarize the following paragraph into 5 bullet points: {para} , Bullet points should be {points}")
para_chain = para_prompt | llm | StrOutputParser()
response = para_chain.invoke({"para": ip_para,"points":points})
print(f"----------------------LLM Output----------------------")
print(response)


----------------------LLM Output----------------------
Here's a summary of the paragraph in 5 bullet points:

*   **Increased Accessibility:** Technology has made high-quality education globally accessible through online platforms.
*   **Enhanced Interaction:** Virtual classrooms enable real-time, barrier-free teacher-student engagement.
*   **Personalized Learning:** AI is customizing education with tutoring, automated grading, and adaptive content.
*   **Student Empowerment:** Learners can now control their pace, review material, and explore beyond traditional curricula.
*   **Challenges:** Digital transformation faces issues of unequal access, isolation, and reduced motivation, requiring a balance with human support and inclusivity.


Question-2: Building Chain using LangChain
You are building a mini content pipeline using an LLM API with LangChain with two chained steps:

Step 1: From the given topic + audience, generate a clear outline.

Step 2: Using only that outline, generate the final content.

Return a strict JSON object with these keys:

outline (array of section titles)
final_content (string, must follow the outline order)
quality_check (string, 1–2 lines confirming the content matches the outline)
Sample input:

topic: “Async/Await in JavaScript”
audience: “Beginners who know callbacks”
length: “~250 words”
tone: “Simple and friendly”
Sample output (JSON):

{
  "outline": [
    "What problem async/await solves",
    "The basic syntax (with a tiny example)",
    "What happens when there is an error (try/catch)",
    "Common beginner mistakes",
    "Quick recap"
  ],
  "final_content": "Async/await helps you write asynchronous code that looks and feels like normal step-by-step code. It mainly solves the problem of messy nested callbacks or long `.then()` chains that are hard to read...\n\nThe basic idea is: you mark a function as `async`, and inside it you can use `await` before a promise to pause until it finishes...\n\nWhen something goes wrong (like a failed API call), you handle it using `try/catch`...\n\nBeginners often forget that `await` works only inside an `async` function, or they assume `await` makes code run faster (it doesn’t—it just makes it easier to read)...\n\nRecap: use `async` to enable `await`, use `await` to wait for promises, and wrap risky code in `try/catch` for errors.",
  "quality_check": "Content follows the outline in the same order and stays beginner-friendly with a simple explanation and common mistakes."
}

In [7]:
!pip install langchain langchain-google-genai

In [8]:
#Problem-2
from langchain_google_genai import ChatGoogleGenerativeAI
from google.colab import userdata
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
llm = ChatGoogleGenerativeAI(model="gemini-2.5-flash-lite",google_api_key=userdata.get('GEMINI_API_KEY'))
topic = "Async/Await in JavaScript"
audience = "Beginners who know callbacks"
length = "~250 words"
tone = "Simple and friendly"
keys = [
    'Outline''(array of section title)',
    'Final_Content''(string, must follow the outline order)',
    'Quality_Check''(1-2 lines confirming content matches outline)'
]
final_prompt = PromptTemplate(input_variables=["topic","audience","length","tone","keys"],template="generate a clear outline based on {topic} + {audience} only using the generated outline generate the Final Content of size:{length} & {tone} generate a JSON object with keys: {keys}")
final_chain = final_prompt | llm | StrOutputParser()
response = final_chain.invoke({"topic":topic,"audience":audience,"length":length,"tone": tone,"keys":keys})
print(f"----------------------FINAL CONTENT----------------------")
print(response)



----------------------FINAL CONTENT----------------------
Here's the outline, final content, and JSON object as requested:

## Outline:

1.  **Introduction: The Callback Challenge**
2.  **Introducing Async/Await: A Smoother Way**
3.  **How Async/Await Works: The Magic Behind the Scenes**
4.  **Benefits for Beginners**
5.  **Simple Example**
6.  **Conclusion**

## Final Content:

Hey there! If you're comfortable with callbacks in JavaScript, you've already taken a big step in handling asynchronous operations. But sometimes, managing nested callbacks, often called "callback hell," can get a bit messy and hard to read.

That's where `async/await` comes in, making your asynchronous code look and feel much more like regular, synchronous code. It’s like a special syntax built on top of Promises that simplifies how you write and understand asynchronous JavaScript.

Under the hood, `async` and `await` work with Promises. The `async` keyword before a function tells JavaScript that this function

Question-3: BMI Calculator using LLM
Write a program that takes a user's name, height (in meters), and weight (in kilograms) as input, and uses Gemini (via LangChain) to calculate the BMI and explain result (underweight, normal, overweight, obese) Use PromptTemplate to format the prompt.

📥 Sample Input
Name: Rahul
Height: 1.55
Weight: 68
You should NOT calculate the BMI yourself.

Instead, pass the inputs to Gemini LLM using a properly formatted prompt.

Gemini LLM should:

Calculate the BMI using the formula: BMI = weight / (height × height)

Return a friendly explanation of what the BMI number means (e.g., underweight, healthy, overweight).



In [10]:
from langchain_google_genai import ChatGoogleGenerativeAI
from google.colab import userdata
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
llm = ChatGoogleGenerativeAI(model="gemini-2.5-flash-lite",google_api_key=userdata.get('GEMINI_API_KEY'))
name = input("Enter Your Name: ")
height = float(input("Enter Your Height (in meters): "))
weight = float(input("Enter Your Weight (in KG): "))
bmi = weight / (height*height)
prompt = PromptTemplate(input_variables=["name","height","weight","bmi"],template="based on th {name} {height} {weight} calculate the {bmi} . absed on tha {bmi} explain the (underweight,normal,overweight,obese). Return a friendly explanation of what the BMI number means (e.g., underweight, healthy, overweight).")
prompt_chain = prompt | llm | StrOutputParser()
response = prompt_chain.invoke({"name":name,"height":height,"weight":weight,"bmi":bmi})
print("----------------------------BMI CALCULATION----------------------")
print(response)

Enter Your Name: Rahul
Enter Your Height (in meters): 1.55
Enter Your Weight (in KG): 68
----------------------------BMI CALCULATION----------------------
Let's break down Rahul's information!

**1. Calculating Rahul's BMI**

To calculate BMI, we use the formula:

BMI = weight (kg) / (height (m))^2

*   **Weight:** 68.0 kg
*   **Height:** 1.55 m

First, we need to square Rahul's height:
1.55 m * 1.55 m = 2.4025 m^2

Now, let's calculate the BMI:
BMI = 68.0 kg / 2.4025 m^2 = **28.303850156087407**

This matches the number you provided!

**2. Explaining Rahul's BMI (28.30)**

Rahul's BMI is approximately **28.3**. Based on standard BMI categories, this falls into the **overweight** range.

**Friendly Explanation of What BMI Means:**

Think of your BMI as a general indicator of your body's weight in relation to your height. It's a quick way to see if you're in a range that's generally considered healthy, or if you might be carrying a bit more or less weight than what's typically associate